pallets/support/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+pub use pallet::*;
+
+pub const MAX_STACK_SIZE: u32 = 16;
+
+const LOG_TARGET: &str = "runtime::amm-support";
+
+type ExecutionIdStack = BoundedVec<ExecutionType, ConstU32<MAX_STACK_SIZE>>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::storage]
+	/// Next available incremental ID
+	#[pallet::getter(fn incremental_id)]
+	pub(super) type IncrementalId<T: Config> = StorageValue<_, IncrementalIdType, ValueQuery>;
+
+	#[pallet::storage]
+	/// Execution context to figure out where the trade is originated from
+	#[pallet::getter(fn execution_context)]
+	pub(super) type ExecutionContext<T: Config> = StorageValue<_, ExecutionIdStack, ValueQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Trade executed.
+		Swapped {
+			swapper: T::AccountId,
+			filler: T::AccountId,
+			filler_type: Filler,
+			operation: TradeOperation,
+			inputs: Vec<Asset>,
+			outputs: Vec<Asset>,
+			fees: Vec<Fee<T::AccountId>>,
+			operation_stack: Vec<ExecutionType>,
+		},
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			ExecutionContext::<T>::kill();
+			
+			T::DbWeight::get().reads_writes(1, 1)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn deposit_trade_event(
+		swapper: T::AccountId,
+		filler: T::AccountId,
+		filler_type: Filler,
+		operation: TradeOperation,
+		inputs: Vec<Asset>,
+		outputs: Vec<Asset>,
+		fees: Vec<Fee<T::AccountId>>,
+	) {
+		let operation_stack = Self::get_context();
+		Self::deposit_event(Event::<T>::Swapped {
+			swapper,
+			filler,
+			filler_type,
+			operation,
+			inputs,
+			outputs,
+			fees,
+			operation_stack,
+		});
+	}
+
+	pub fn add_to_context<F>(execution_type: F) -> Result<IncrementalIdType, DispatchError>
+	where
+		F: FnOnce(u32) -> ExecutionType,
+	{
+		let next_id = IncrementalId::<T>::try_mutate(|current_id| -> Result<IncrementalIdType, DispatchError> {
+			let inc_id = *current_id;
+			*current_id = current_id.overflowing_add(1).0;
+			Ok(inc_id)
+		})?;
+
+		ExecutionContext::<T>::try_mutate(|stack| -> DispatchResult {
+			//We make it fire and forget, and it should fail only in test and when if wrongly used
+			debug_assert_ne!(stack.len(), MAX_STACK_SIZE as usize, "Stack should not be full");
+			if let Err(err) = stack.try_push(execution_type(next_id)) {
```

Comment:
There might be an issue with this. Imagine a scenario where we push 15 times to the stack. The stack is populated 10 times, and ignored 5 times. Then let's reduce the stack 7 times. Now the stack has 3 items, but it should have 8 items.